src/include/txn.i

```diff
@@ -193,6 +193,176 @@ __wt_timestamp_subone(wt_timestamp_t *ts)
 #endif /* HAVE_TIMESTAMPS */
 
+/*
+ * __wt_txn_op_set_recno --
+ *      Set the latest transaction operation with the given recno.
+ */
+static inline void
+__wt_txn_op_set_recno(WT_SESSION_IMPL *session, uint64_t recno)
+{
+	WT_TXN *txn;
+	WT_TXN_OP *op;
+
+	txn = &session->txn;
+
+	WT_ASSERT(session, txn->mod_count > 0 && recno != WT_RECNO_OOB);
+	op = txn->mod + txn->mod_count - 1;
+
+	WT_ASSERT(session, op->type == WT_TXN_OP_BASIC_COL ||
+	    op->type == WT_TXN_OP_INMEM_COL);
+
+	op->u.op_col.recno = recno;
+}
+
+/*
+ * __txn_resolve_prepared_update --
+ *      Resolve a prepared update as committed update.
+ */
+static inline void
+__txn_resolve_prepared_update(WT_SESSION_IMPL *session, WT_UPDATE *upd)
+{
+#ifdef HAVE_TIMESTAMPS
+	WT_TXN *txn;
+
+	txn = &session->txn;
+	/*
+	 * In case of a prepared transaction, the order of modification of the
+	 * prepare timestamp to commit timestamp in the update chain will not
+	 * affect the data visibility, a reader will encounter a prepared
+	 * update resulting in prepare conflict.
+	 *
+	 * As updating timestamp might not be an atomic operation, we will
+	 * manage using state.
+	 */
+	upd->prepare_state = WT_PREPARE_LOCKED;
+	WT_WRITE_BARRIER();
+	__wt_timestamp_set(&upd->timestamp, &txn->commit_timestamp);
+	WT_PUBLISH(upd->prepare_state, WT_PREPARE_RESOLVED);
+#else
+	WT_UNUSED(session);
+	WT_UNUSED(upd);
+#endif
+}
+
+/*
+ * __txn_prepared_op_resolve --
+ *      Resolve a transaction operation indirect references.
+ *
+ *      In case of prepared transactions, the prepared updates could be evicted
+ *      using cache overflow mechanism. Transaction operations referring these
+ *      prepared updates would be referring to them using indirect references
+ *      (i.e keys/recnos), which need to be resolved as part of that
+ *      transaction commit/rollback.
+ */
+static inline int
+__txn_prepared_op_resolve(WT_SESSION_IMPL *session, WT_TXN_OP *op, bool commit)
+{
+	WT_CURSOR *cursor;
+	WT_DECL_RET;
+	WT_TXN *txn;
+	WT_UPDATE *upd;
+	const char *open_cursor_cfg[] = {
+	    WT_CONFIG_BASE(session, WT_SESSION_open_cursor), NULL };
+
+	txn = &session->txn;
+
+	if (op->type == WT_TXN_OP_NONE || op->type == WT_TXN_OP_REF_DELETE ||
+	    op->type == WT_TXN_OP_TRUNCATE_COL ||
+	    op->type == WT_TXN_OP_TRUNCATE_ROW)
+		return (0);
+
+	WT_RET(__wt_open_cursor(session,
+	    op->btree->dhandle->name, NULL, open_cursor_cfg, &cursor));
+
+	/*
+	 * Transaction prepare is cleared temporarily as cursor functions are
+	 * not allowed for prepared transactions.
+	 */
+	F_CLR(txn, WT_TXN_PREPARE);
+	if (op->type == WT_TXN_OP_BASIC_ROW ||
+	    op->type == WT_TXN_OP_INMEM_ROW)
+		__wt_cursor_set_raw_key(cursor, &op->u.op_row.key);
+	else
+		((WT_CURSOR_BTREE *)cursor)->iface.recno =
+		    op->u.op_col.recno;
+	F_SET(txn, WT_TXN_PREPARE);
+
+	WT_WITH_BTREE(session,
+	    op->btree, ret = __wt_btcur_search_uncommitted(
+	    (WT_CURSOR_BTREE *)cursor, &upd));
+	WT_ERR(ret);
+	WT_ASSERT(session, upd != NULL);
```

Comment:
I think this is going away, but it's also duplicated inside __wt_btcur_search_uncommitted - choose one or the other.